---
title: "GSI_module_development"
author: "Matthew Bootsma"
date: "May 24, 2019"
output: html_document
---
This module will need two dependent inputs, 
1) The formatted rubias dataframe, produced using vcf_2_rubias.Rmd
2) An associated table of metadata for each SNP/micro-haplotype that will allow you to select markers based on diversity statistics, e.g., FST or mhap_He
      This file can be produced using SNP_summary_table.Rmd
#Dependencies
```{r}
'%!in%' <- function(x,y)!('%in%'(x,y))
library(dplyr)
```
#Import metadata
```{r}
selection_dat = read_excel("./data/panel_data-ALL/SNP_selection_table_v3.1.xlsx",sheet = "SNP_selection_table_v3")
```
#Select markers here!
Marker selection is based on metadata calculated in SNP_summary_table.Rmd

Every panel test will require selecting different markers 
Given my results showed FST was ideal for GSI and mhap_HE was ideal for parentage this is what the decision will be based on
You can call different variables if you want, just edit the code. I'm looking to make this a function but have yet to do so.

NOTE: start by indicating the number of loci you want to select from each catagory. 
Given that some of the top FST markers will also be the top mhap_HE markers, I select FST markers, exclude them from the list, then select mhap_HE markers. 
My results indicated this was giving better accuracy than the other way around.
```{r}
#DON'T FORGET TO NAME PANEL!!!
panel_name = "TEST_NAME"
N_LOCI_IN_PANEL = 600
N_FST_LOCI = 350
#The number of mhap_HE loci will be the difference between N_LOCI_IN_PANEL and N_FST_LOCI
#Apply cutoff for markers with extreme FIS values
selection_dat_FIS = selection_dat[which(selection_dat$`Fis-diffCalc` > -0.2 & selection_dat$`Fis-diffCalc` < 0.2),]
#Filter out any markers with position <17 for primer design
#These will still be in the sequence data you use for primer design, so we can design with wobble positions if appropriate
selection_dat_FIS = selection_dat_FIS[which(selection_dat_FIS$POS > 16),]
#Grab the FST loci (good for GSI)
selection_dat_FST = arrange(selection_dat_FIS,desc(`Fst-diffCalc`))
LOCI_FST_only = unique(selection_dat_FST$LOCUS_ID)[1:N_FST_LOCI]
#Remove loci in the FST panel from the data so we don't select a locus twice
selection_dat_parentage = arrange(selection_dat_FIS,desc(`HAP_He-adegenet-summary`))
selection_dat_parentage = selection_dat_parentage[which(selection_dat_parentage$LOCUS_ID %!in% LOCI_FST_only),]
LOCI_parentage = unique(selection_dat_parentage$LOCUS_ID)[1:(N_LOCI_IN_PANEL-n_distinct(LOCI_FST_only))]
#GREAT, now we can merge them to get 600 loci and give it a test
loci_for_panel = c(LOCI_FST_only,LOCI_parentage)
#Now saving the panel for documentation and reproduction
#First make a directory to save results, if it already exists this line will do nothing so no worries.
dir.create("./putative_panel")

out_file = paste("./putative_panel/",panel_name,".rda", sep = "")
save(loci_for_panel, file = out_file)
```

